### PR TITLE
Use http based promise in emulated trait and specs

### LINKED ISSUE #34: Clear exception usage in Plugins
| Q            | A
| ------------ | ---
| Bug?         | yes
| New Feature? | no
| Version      | 1.1-

#### Actual Behavior

Plugin implementations only accept `Http\Client\Exception` instances in their rejection handler ([here](https://github.com/php-http/client-common/blob/master/src/Plugin/HistoryPlugin.php#L43) and [here](https://github.com/php-http/client-common/blob/master/src/Plugin/RetryPlugin.php#L65)) while there is no such limitation in the promise spec or the [RejectedPromise](https://github.com/php-http/promise/blob/master/src/RejectedPromise.php#L20).

#### Expected Behavior

Plugins should accept any kind of exceptions and handle only which are acceptable for them and rethrow the rest. This is not a breaking change since we just need to widen the accepted exception types in the plugins.

#### Steps to Reproduce

See https://github.com/php-http/client-common/pull/33

#### Possible Solutions
As said, widen the range of accepted exception types.

### PR DESCRIPTION
**Pull request by joelwurtz**
| Q               | A
| --------------- | ---
| Bug fix?        | yes
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | fixes #34 , replace #35
| Documentation   | /
| License         | MIT

See issue #34 and discussion in #35 also for more information

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -12,7 +12,7 @@
     ],
     "require": {
         "php": ">=5.4",
-        "php-http/httplug": "^1.0",
+        "php-http/httplug": "^1.1.0",
```
Comment (GrahamCampbell) [inline composer.json]: `^1.1` would be fine here

Comment (joelwurtz) [inline composer.json]: Yes, it's better, thanks.

**Comment by GrahamCampbell** [comment]
Ok, so what happens when a non-http exception gets thrown?

**Comment by Nyholm** [comment]
When a non-httplug exception is thrown we abort the plugin chain and the exception bubble up like normal. Only httplug exception will respect the plugin chain.

**Comment by GrahamCampbell** [comment]
This will break other libraries though.

**Comment by sagikazarmark** [comment]
I don't think so. If you used a plugin with rejection callbacks broke your library too. This was a bug IMO, which has been solved. Also, this bug only affects you if you use plugins.

File: src/Plugin.php [hunk 2]
```diff
@@ -24,7 +24,7 @@
      * @param callable         $next    Next middleware in the chain, the request is passed as the first argument
      * @param callable         $first   First middleware in the chain, used to to restart a request
      *
-     * @return Promise
+     * @return Promise Resolves a PSR-7 Response or fails with an Http\Client\Exception (The same as HttpAsyncClient).
```
Comment (joelwurtz) [inline src/Plugin.php]: Is this a BC break ? (For me no, it's more a clarification but i dont know if this interface is already used but other packages than one we have under control ?)

**Comment by sagikazarmark** [comment]
I added some explanation here: https://github.com/php-http/httplug/releases/tag/v1.1.0

File: src/Plugin.php (hunk 2 above)
Comment (Nyholm) [inline src/Plugin.php]: Could we improve this doc block somehow even more?

Maybe describe what the callables will return?

Comment (joelwurtz) [inline src/Plugin.php]: Don't we have already this on the documentation ? if so would just add a link with `@see`

Comment (Nyholm) [inline src/Plugin.php]: We don't but we should. Write 
`@see http://php-http.readthedocs.io/en/latest/plugins/introduction.html`

FYI: https://github.com/php-http/documentation/issues/149

**Comment by Nyholm** [comment]
Thank you @joelwurtz. 

Is this ready for merge?

File: src/Plugin.php (hunk 2 above)
Comment (sagikazarmark) [inline src/Plugin.php]: From my POV, it's not. Http Promises still fullfill the Promise interface.

**Comment by sagikazarmark** [comment]
Please gimme a note when this can be merged and tagged because I would like to do some other things too.

File: src/Plugin.php (hunk 2 above)
Comment (Nyholm) [inline src/Plugin.php]: Add "@see http://docs.php-http.org/en/latest/plugins/build-your-own.html". I've added some lines about exceptions (https://github.com/php-http/documentation/pull/151)

> Is this a BC break ?

No, this is not a BC break.

**Comment by joelwurtz** [comment]
For me all the work is done here, so feel free to merge if it's ok to you (not a fan of merging my own PR in a collaboration project, it feels presumptuous...)

**Comment by Nyholm** [comment]
Thank you for this PR. Good work!

**Pull request merged**